### ISSUE TITLE
Positional audio: round table-like arrangement

### ISSUE THREAD

**Issue by bergfried** [mumble-voip/mumble#6308]
### Context

Using Mumble as conference software.

### Description

I would like to see a positional audio plugin included in Mumble where all participants are placed as if sitting at a round table in a conference room.

That is, every user selecting it as the plugin for positional audio will be positioned at the outline of an imaginary circle with a fixed minimum distance between each other user, facing the center of the circle. Assuming alphabetical order (more specifically, the order in which users are listed in the channel overview), every user should hear the "next" user to her left, end the "previous" user to her right, implying that all user's positions are laid out clockwise in ascending alphabetical order.

This might make Mumble more appealing for serious applications like online conferences, since listening to different people from different angles might make it easier to focus on one of them (think cocktail party effect).

For demonstration purposes, please have a look at the following Python script that calculates the numbers you need to enter into the "Manual placement" plugin to get an idea of the final result.

```
#!/usr/bin/env python3

import sys
import math

def radius_2d(min_distance, number_of_participants):
    return min_distance / 2 / math.sin(math.pi / number_of_participants)

def ideal_position_data_2d(min_distance, number_of_participants):
    r = radius_2d(min_distance, number_of_participants)
    for i in range(number_of_participants):
        alpha = i / number_of_participants * math.tau
        yield (r * math.cos(alpha), 0.0, r * math.sin(alpha), math.degrees((-alpha + 0.5 * math.pi) % math.tau), 0.0)

def main(args):
    min_distance = float(args[1])
    context = args[2]
    participants = args[3:]
    number_of_participants = len(participants)
    for (part, pos) in zip(participants, ideal_position_data_2d(min_distance, number_of_participants)):
        print("%15s:  X=% .2f  Y=% .2f  Z=% .2f  azimuth=%3d  elevation=%d  context=%s  identity=%s" % \
            (part, round(pos[0], 2), round(pos[1], 2), round(pos[2], 2), round(pos[3], 0), round(pos[4], 0), context, part))

if __name__ == "__main__":
    main(sys.argv)
```

For example, if you call it like this …

```
./demo.py 1.5 rdtbl Alice Bob Chuck
```

… and Alice, Bob and Chuck enter the numbers as computed for each of them, ideally, they should hear each other as if sitting at a round table, with at least 1.5 meters between each of them.

Bonus points if priority speakers are placed as if floating, talking down from above the center of the imaginary table.

(Actually, I am not sure whether I got the computation right, but it should be enough to get the idea.)

### Mumble component

Both

### OS-specific?

No

### Additional information

Note that I did not try to write such a dedicated plugin myself for various reasons, but especially because I cannot remember the plugin API exposing all of the relevant data (especially the list of all user's names in the same context) for this kind of plugin to be be written as a "normal" positional audio plugin. Besides, having such a plugin built into every Mumble client by default would make adoption much better, maybe even giving Mumble a feature to distinguish it from other, less privacy-oriented conference software.

**Comment by Krzmbrzl** [comment]
More generally, this kind of thing would probably also benefit of introducing a local-only positional audio where not every client has to enable it in order for things to work. Instead, users could simply place other clients spatially as desired and have positional audio work regardless of whether or not the other participants use it.
E.g. like the manual plugin but instead of only placing yourself, you place everyone.

Arranging folks equidistantly around a circle would then simply be a convenience feature.

One would have to ensure however, that folks won't end up being muted by positional audio (at least not accidentally) as this is likely not what would be desired in such non-gaming related scenarios.

**Comment by Akasen** [comment]
So I'm gonna put my two cents on this and say that this would be an extremely interesting feature for something like playing TTRPG's online. 

I think a feature like this, or just an all around overhaul to the manual position plugin and its documentation, would be appreciated.

**Comment by saiprakash1807** [comment]
Hi, I am a beginner and would love to contribute to this feature!

I have gone through the issue and the provided Python script, and I understand the requirement of arranging users in a virtual round-table fashion using positional audio.

I would like to start working on this — initially by exploring the Mumble codebase to check for:
- Access to the current user list in a channel.
- How positional audio data is applied to users.
- How plugins or built-in features can be integrated.

Let me know if this is okay and if there are any specific guidelines or files I should look into to get started.

Thanks!

**Comment by Krzmbrzl** [comment]
Contributions are always very welcome! I'll try to give some more details in a couple of days (I'm a bit short of time as of right now). Should I forget, feel free to ping me and remind me that I still owe you an explanation.

**Comment by Pastoray** [comment]
> Contributions are always very welcome! I'll try to give some more details in a couple of days (I'm a bit short of time as of right now). Should I forget, feel free to ping me and remind me that I still owe you an explanation.

@Krzmbrzl I am currently looking into this and i would like to know more about this especially considering the already existing "Manual Positional Audio" plugin is already very similar to this request, so am assuming the best course of action would be an extension of that plugin ?

**Comment by Krzmbrzl** [comment]
> so am assuming the best course of action would be an extension of that plugin ?

yeah, that would likely be the most straight-forward way of achieving this. In this context, we would likely want to have functionality in the manual plugin that assigns the local user a position inside a circle-arrangement of users. For this, we need a way to determine how many people will be part of the "round table" and then we need a way to uniquely order all participants in that group (we could simply use their session IDs for that purpose as those are assigned by the server and hence all clients have the same information about who has which session ID). Then we need to determine the radius of the table (e.g. `2m`) and then compute the local user's position on that table. For that, determine the position of the local user in the group (simply order all group participants by session ID and then check the local user's index in that sorted list) and compute their position as
```
x = 2 * cos((360° / <size of group>) * <user index>)
y = 2 * sin((360° / <size of group>) * <user index>)
```
These are the coordinates that we have to enter into the manual plugin. If everyone of that group does this, they will effectively have created the desired round-table-like effect.

The interesting questions are
- how to determine who is part of the group? The easiest choice would be to take everyone in the local user's channel
- handling dynamic adding and removing participants from the group (e.g. because they are entering or leaving the local user's channel). Every time this happens, we have to recompute the local user's position.

Let me know if you have further questions

**Comment by Pastoray** [comment]
> > so am assuming the best course of action would be an extension of that plugin ?
> 
> yeah, that would likely be the most straight-forward way of achieving this. In this context, we would likely want to have functionality in the manual plugin that assigns the local user a position inside a circle-arrangement of users. For this, we need a way to determine how many people will be part of the "round table" and then we need a way to uniquely order all participants in that group (we could simply use their session IDs for that purpose as those are assigned by the server and hence all clients have the same information about who has which session ID). Then we need to determine the radius of the table (e.g. `2m`) and then compute the local user's position on that table. For that, determine the position of the local user in the group (simply order all group participants by session ID and then check the local user's index in that sorted list) and compute their position as
> 
> ```
> x = 2 * cos((360° / <size of group>) * <user index>)
> y = 2 * sin((360° / <size of group>) * <user index>)
> ```
> 
> These are the coordinates that we have to enter into the manual plugin. If everyone of that group does this, they will effectively have created the desired round-table-like effect.
> 
> The interesting questions are
> 
>     * how to determine who is part of the group? The easiest choice would be to take everyone in the local user's channel
> 
>     * handling dynamic adding and removing participants from the group (e.g. because they are entering or leaving the local user's channel). Every time this happens, we have to recompute the local user's position.
> 
> 
> Let me know if you have further questions

I've been trying to get the "manual placement" plugin to work to better understand what you already have working, i couldn't get it to work, using two accounts from the same PC, i've tried to set their context to the same string, with different and identical identity, neither actually seemed to work ? i couldn't figure out how to get a second dot on the plane (my alt on the same channel), and every time i go back to verify that the two have the same context it's actually reset back to ("Mumble", "Agent47").

If you have the time i'd appreciate a bit of an explanation about the plugin.
i'd also like to know if the already existing plugin is game specific (aka only works when you share the context of the game).

**Comment by datietou** [comment]
已收到您的邮件，我将尽快回复，谢谢！
